File: OJ/leet377/leet377.cpp
/*
leet377


*/


#include "../utils/LeetCpp.utils.hpp"
using namespace std;

class Solution {
public:
    int combinationSum4(const vector<int>& A, int target) {
        // 完全背包, 有序
        unsigned long long f[1005]{1};
        for (int t = 0; t <= target; ++t) {
            for (auto&& x: A) {
                if (t >= x) f[t] += f[t - x];
            }
        }
        return f[target];
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.combinationSum4( {1,2,3}, 4 ) << endl;
    cout << sol.combinationSum4( {3,2,1}, 4 ) << endl;
    cout << sol.combinationSum4( {1,3,2}, 4 ) << endl;
    cout << sol.combinationSum4( {3,1,2}, 4 ) << endl;
    cout << sol.combinationSum4( {2,3,1}, 4 ) << endl;
    cout << sol.combinationSum4( {2,1,3}, 4 ) << endl;
    return 0;
}

File: OJ/leet377/leet377.py
# leet377

from typing import List


class Solution:
    def combinationSum4(self, Ws: List[int], target: int) -> int:
        # f(v) = sum f(v - wi) for wi in Ws 
        # 由于有顺序, f(v) 表示这 N 个物品能够填充的方法数目
        # 对于容量 v 的包裹，第一个位置放入 wi, 则剩余开放有 f(v-wi)
        n = len(Ws)
        f = [0] * (target+1)
        f[0] = 1
        for v in range(1, target+1):
            for wi in Ws:
                if v >= wi:
                    f[v] += f[v-wi]
        return f[target]



print(Solution().combinationSum4([1,2,3], 4))


